File: pkg/client/pki.go
package client

import "context"

// PKIClient manages the PKI secrets engine.
type PKIClient struct{ c *Client }

// PKI returns a client for the PKI secrets engine.
func (c *Client) PKI() *PKIClient { return &PKIClient{c: c} }

// PKICARequest configures a new CA to generate.
type PKICARequest struct {
	CommonName   string   `json:"common_name"`
	TTL          string   `json:"ttl,omitempty"`          // e.g. "87600h"
	KeyType      string   `json:"key_type,omitempty"`     // "ec" (default) or "rsa"
	KeyBits      int      `json:"key_bits,omitempty"`     // 256, 384, 521 for EC; 2048, 4096 for RSA
	Organization []string `json:"organization,omitempty"`
	Country      []string `json:"country,omitempty"`
}

// PKIRole defines the certificate issuance policy for a named role.
type PKIRole struct {
	Name            string   `json:"name"`
	AllowedDomains  []string `json:"allowed_domains,omitempty"`
	AllowSubdomains bool     `json:"allow_subdomains,omitempty"`
	AllowAnyDomain  bool     `json:"allow_any_domain,omitempty"`
	MaxTTL          string   `json:"max_ttl,omitempty"`
	KeyType         string   `json:"key_type,omitempty"`
	KeyBits         int      `json:"key_bits,omitempty"`
	Organization    []string `json:"organization,omitempty"`
	Country         []string `json:"country,omitempty"`
}

// PKIIssueRequest specifies parameters for issuing a certificate.
type PKIIssueRequest struct {
	CommonName string   `json:"common_name"`
	AltNames   []string `json:"alt_names,omitempty"`
	IPSANs     []string `json:"ip_sans,omitempty"`
	TTL        string   `json:"ttl,omitempty"`
}

// PKICert is a certificate returned by issue or get operations.
type PKICert struct {
	Certificate string `json:"certificate"`
	PrivateKey  string `json:"private_key,omitempty"`
	SerialNumber string `json:"serial_number"`
	CAChain     string `json:"ca_chain,omitempty"`
	Expiration  int64  `json:"expiration,omitempty"`
}

// GenerateCA initialises the PKI engine with a new self-signed CA.
// Returns the CA certificate PEM.
func (p *PKIClient) GenerateCA(ctx context.Context, req PKICARequest) (string, error) {
	var resp struct {
		Certificate string `json:"certificate"`
	}
	if err := p.c.post(ctx, "/v1/pki/generate/root", req, &resp); err != nil {
		return "", err
	}
	return resp.Certificate, nil
}

// GetCACert returns the current CA certificate PEM.
func (p *PKIClient) GetCACert(ctx context.Context) (string, error) {
	var resp struct {
		Certificate string `json:"certificate"`
	}
	if err := p.c.get(ctx, "/v1/pki/ca/pem", &resp); err != nil {
		return "", err
	}
	return resp.Certificate, nil
}

// PutRole creates or updates a PKI role.
func (p *PKIClient) PutRole(ctx context.Context, role PKIRole) error {
	return p.c.putJSON(ctx, "/v1/pki/roles/"+role.Name, role)
}

// GetRole reads a PKI role.
func (p *PKIClient) GetRole(ctx context.Context, name string) (*PKIRole, error) {
	var out PKIRole
	if err := p.c.get(ctx, "/v1/pki/roles/"+name, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// DeleteRole removes a PKI role.
func (p *PKIClient) DeleteRole(ctx context.Context, name string) error {
	return p.c.doDelete(ctx, "/v1/pki/roles/"+name)
}

// IssueCert issues a certificate for the given role.
func (p *PKIClient) IssueCert(ctx context.Context, role string, req PKIIssueRequest) (*PKICert, error) {
	var out PKICert
	if err := p.c.post(ctx, "/v1/pki/issue/"+role, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RevokeCert revokes a certificate by serial number.
func (p *PKIClient) RevokeCert(ctx context.Context, serial string) error {
	return p.c.post(ctx, "/v1/pki/revoke/"+serial, nil, nil)
}

// GetCert reads a certificate by serial number.
func (p *PKIClient) GetCert(ctx context.Context, serial string) (*PKICert, error) {
	var out PKICert
	if err := p.c.get(ctx, "/v1/pki/certs/"+serial, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: web/src/shared/api/endpoints/pki.ts
import { instance } from '../axios'
import { listRequest } from '../list'

export interface PKIRole {
    name: string
    allowed_domains?: string[]
    allow_subdomains?: boolean
    allow_ip_sans?: boolean
    allow_localhost?: boolean
    key_type?: 'ec' | 'rsa'
    key_bits?: number
    default_ttl?: number
    max_ttl?: number
    server_flag?: boolean
    client_flag?: boolean
}

export interface PKIRolePut {
    allowed_domains?: string[]
    allow_subdomains?: boolean
    key_type?: 'ec' | 'rsa'
    default_ttl?: string
    max_ttl?: string
    server_flag?: boolean
    client_flag?: boolean
}

export interface IssuedCert {
    serial: string
    certificate: string
    private_key: string
    issuing_ca: string
    expires_at: string
    ttl: number
}

export interface CertRecord {
    serial: string
    role_name: string
    common_name: string
    sans?: string[]
    issued_at: string
    expires_at: string
    revoked: boolean
    revoked_at?: string
}

export async function generateRootCA(body: {
    common_name: string
    ttl?: string
    key_type?: 'ec' | 'rsa'
}): Promise<{ certificate: string }> {
    const { data } = await instance.post<{ certificate: string }>('/v1/pki/generate/root', body)
    return data
}

export async function importCA(cert_pem: string, key_pem: string): Promise<void> {
    await instance.post('/v1/pki/import/ca', { cert_pem, key_pem })
}

export async function listPKIRoles(): Promise<string[]> {
    return listRequest('/v1/pki/roles/')
}

export async function putPKIRole(name: string, body: PKIRolePut): Promise<PKIRole> {
    const { data } = await instance.put<PKIRole>(`/v1/pki/roles/${name}`, body)
    return data
}

export async function deletePKIRole(name: string): Promise<void> {
    await instance.delete(`/v1/pki/roles/${name}`)
}

export async function issueCert(
    role: string,
    body: { common_name: string; alt_names?: string[]; ttl?: string }
): Promise<IssuedCert> {
    const { data } = await instance.post<IssuedCert>(`/v1/pki/issue/${role}`, body)
    return data
}

export async function revokeCert(serial: string): Promise<void> {
    await instance.post(`/v1/pki/revoke/${serial}`, {})
}

export async function listCerts(): Promise<string[]> {
    return listRequest('/v1/pki/certs/')
}

export async function getCert(serial: string): Promise<CertRecord> {
    const { data } = await instance.get<CertRecord>(`/v1/pki/certs/${serial}`)
    return data
}
